followthemoney/proxy.py

```diff
@@ -79,18 +79,16 @@ def __init__(
         #: than ``id``, ``schema`` or ``properties``, they will be kept in here
         #: and re-added upon serialization.
         self.context = data
-        self._properties: Dict[str, Set[str]] = {}
+        self._properties: Dict[str, List[str]] = {}
         self._size = 0
 
-        for key, value in properties.items():
+        for key, values in properties.items():
             if key not in self.schema.properties:
                 continue
-            if not cleaned:
-                self.add(key, value, cleaned=cleaned, quiet=True)
+            if cleaned:
+                self.add(key, values, cleaned=cleaned)
             else:
-                values = set(value)
-                self._properties[key] = values
-                self._size += sum([len(v) for v in values])
+                self.add(key, values, quiet=True)
```

Comment:
This was previously setting `_properties` directly in case the data is already cleaned. For the sake of simplicity I’m calling the `add` here too. Not sure if this has any disadvantages? (I assume the previous implementation was to avoid unnecessary method calling/loop overhead?)

Reply:
Yep this is all super perf optimised. I'd sort of consider keeping it unrolled, this gets called millions and millions of times e.g. when a mapping is generated and saving three lines in exchange for a few hours on the other end may be bad math.

Reply:
Yes, makes sense, I’ll change that :)

Reply:
I pushed another commit that inlines this again and ran a very basic benchmark based on the `contrib/perf_create_proxy.py` script, creating 2,000,000 proxies with cleaned values. Surprisingly, the inlined version is ~1,4 times slower than the "unoptimized" version (which is only slightly slower than the current FtM release).

Might need to run a slightly more advanced benchmark though to exclude other factors as the reason for this.

### https://github.com/alephdata/followthemoney/pull/1148/commits/f61f7395355b50b83f2cd47f5be471607c4d41a1
Took 10.915786981582642
Took 10.810703992843628
Took 10.882683038711548
Took 10.802056074142456
Took 10.878925085067749

### https://github.com/alephdata/followthemoney/pull/1148/commits/00bb7918169dffa10027a3348ab16d3a7b8eac2e
Took 15.575160026550293
Took 15.75855827331543
Took 15.749438047409058
Took 15.530592918395996
Took 15.721524953842163

### Current followthemoney release
Took 10.602813005447388
Took 10.623345851898193
Took 10.726281881332397
Took 10.77332091331482
Took 10.679194211959839

Reply:
Wooo, weird. I guess we technically don't need to `set` but it feels a bit hygienic....

Reply:
> Wooo, weird. I guess we technically don't need to set but it feels a bit hygienic....

You mean we wouldn’t need to dedupe the values if `clenaed=True` because we’d expect the consumer to already provide deduplicated values? I think I agree in that the API might not officially guarantee that atm, but it’s been like this forever, so it’s a bit risky to change this now…

Reply:
I’m benchmarking a few different scenarios right now, it seems that the results look a bit different when you create a proxy with properties that are already lists for example:

```
{
  "schema": "Person",
  "properties": {
    "name": ["John Doe"],
  },
}
```

```
{
  "schema": "Person",
  "properties": {
    "name": "John Doe",
  }
}
```

Reply:
Okay, ran a slightly more sophisticated benchmark. The results suggest that inlining initialization of property values into the constructor indeed has a significant impact. It’s still slower than the currently released implementation, but it’s less worse (only 25% slowdown vs >100% slowdown).

@pudo The numbers below are for 10x creating 1,000,000 proxies. The slowdown sounds tolerable to me, what do you think?

<img width="1094" alt="Screen Shot 2023-07-11 at 12 50 57" src="https://github.com/alephdata/followthemoney/assets/1512805/328cabe2-6035-4eb9-8a58-31b12f2d9ce9">

I’m not an expert in benchmarking/profiling, so this might still be a little flawed. Ran this in Docker containers for ease of testing different Python versions, but can see similar results when running it on the host machine directly.

Reply:
Yeah all the non-add ones look totally fine :)

Reply:
@pudo Perfect, thanks!

And sorry for bothering you again, but would you mind clarifying what you meant here. I’m not 100% sure I understand what you mean. The set is only used to test whether a value already occurred before, but the values are added to the list of values in their original order.

> Ah but thinking about it a bit more: you can't set there. It's called every time the thing is fished out of the database, which would make this whole PR not persistent....